sample-rust/src/main.rs

```diff
@@ -0,0 +1,92 @@
+use std::{
+    ffi::CString,
+    fs::{File, OpenOptions},
+    io::{prelude::*, Result, SeekFrom},
+    mem::size_of,
+    net::{TcpListener, TcpStream},
+};
+
+fn main() -> Result<()> {
+    let mut file = File::open("/var/log/dpkg.log")?;
+    println!(">>>>> open file {:#?}", file);
+
+    let mut buffer = vec![0; u16::MAX as usize];
+    let read_count = file.read(&mut buffer)?;
+    println!(">>>>> read {:#?} bytes from file {:#?}", read_count, file);
+
+    let new_start = file.seek(SeekFrom::Start(10))?;
+    println!(
+        ">>> seek now starts at {:#?} for file {:#?}",
+        new_start, file
+    );
+
+    let mut write_file = OpenOptions::new()
+        .create(true)
+        .write(true)
+        .read(true)
+        .open("/tmp/rust_sample.txt")
+        .unwrap();
+    println!(">>>>> created file with write permission {:#?}", write_file);
+
+    let written_count = write_file.write("mirrord sample rust".as_bytes()).unwrap();
+    println!(
+        ">>>>> written {:#?} bytes to file {:#?}",
+        written_count, write_file
+    );
+
+    write_file.seek(SeekFrom::Start(0)).unwrap();
+    println!(">>>>> seeking file back to start {:#?}", write_file);
+
+    let mut read_buf = vec![0; 128];
+    let read_count = write_file.read(&mut read_buf).unwrap();
+    println!(
+        ">>>>> read {:#?} bytes from file {:#?}",
+        read_count, write_file
+    );
+
+    let read_str = String::from_utf8_lossy(&read_buf);
+    println!(">>>>> read {:#?} from file {:#?}", read_str, write_file);
+
+    unsafe {
+        let filepath = CString::new("/tmp/rust_sample.txt").unwrap();
+        let file_mode = CString::new("r").unwrap();
+        let file_ptr = libc::fopen(filepath.as_ptr(), file_mode.as_ptr());
```

Comment:
Most straightforward way I found to try out some of the hooked file functions, should I remove it and use only the normal Rust file API?

Reply:
Nope, was just wondering. So you're saying the normal Rust file API doesn't provide full coverage of the hooks?

Reply:
Maybe it does, I could look for the Rust functions that calls some of these `f` prefixed C functions. This was more a case of convenience rather than anything technical.

@aviram actually suggested using a python sample, as it supposedly has a more C-like API, but I'm very unfamiliar with the language, so went for Rust instead.